cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java

```diff
@@ -103,50 +95,29 @@
       RetryStrategies.retryOnConflict(20, 100)
     );
     this.datasetFramework = datasetFramework;
-    this.subscriberThreads = new ArrayList<>();
   }
 
   @Override
-  protected void startUp() throws Exception {
-    taskExecutorService = MoreExecutors.listeningDecorator(
-      Executors.newCachedThreadPool(Threads.createDaemonThreadFactory("notification-subscriber-task")));
-  }
-
-  @Override
-  protected void run() {
+  protected void doStart() {
     LOG.info("Start running NotificationSubscriberService");
-    if (!isRunning()) {
-      return;
-    }
-
-    // TODO: (CDAP-11407) Need to read starting messageId's from a store
-    subscriberThreads.add(new TimeEventNotificationSubscriberThread(null));
-    subscriberThreads.add(new DataEventNotificationSubscriberThread(null));
-
-    for (NotificationSubscriberThread thread : subscriberThreads) {
-      thread.start();
-    }
+    taskExecutorService = MoreExecutors.listeningDecorator(
+      Executors.newCachedThreadPool(new ThreadFactoryBuilder().setNameFormat("constraint-checker-task").build()));
 
-    for (NotificationSubscriberThread thread : subscriberThreads) {
-      Uninterruptibles.joinUninterruptibly(thread);
-    }
+    taskExecutorService.submit(new TimeEventNotificationSubscriberThread());
+    taskExecutorService.submit(new DataEventNotificationSubscriberThread());
   }
 
   @Override
-  protected void triggerShutdown() {
+  protected void doStop() {
     LOG.info("Stopping NotificationSubscriberService.");
-    for (NotificationSubscriberThread thread : subscriberThreads) {
-      thread.interrupt();
-    }
-
-    LOG.info("NotificationSubscriberService stopped.");
-  }
-
-  @Override
-  protected void shutDown() throws Exception {
-    if (taskExecutorService != null) {
-      taskExecutorService.shutdownNow();
+    try {
+      if (!taskExecutorService.awaitTermination(5, TimeUnit.SECONDS)) {
+        taskExecutorService.shutdownNow();
+      }
+    } catch (InterruptedException ie) {
+      Thread.currentThread().interrupt();
```

Comment:
need to shutdownNow(), reagrdless of whether the awaitForTermination() was interrupted.